### PR TITLE
Join city reminder

### PR DESCRIPTION
**Pull request by vovikboyko**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: EPlast/EPlast.BLL/Interfaces/EmailSending/IEmailReminderService.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading.Tasks;
+
+namespace EPlast.BLL.Interfaces
+{
+    public interface IEmailReminderService
+    {
+        Task JoinCityReminder();
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Interfaces/EmailSending/IEmailReminderService.cs:7]:
```suggestion
        Task JoinCityReminderAsync();
```

**Review by AlexKaschuk — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs [hunk 2]
```diff
@@ -161,7 +119,82 @@ public async Task ChangeCurrentRoleAsync(string userId, string role)
...
+                }).ToList();
+            }
+
+            var citiesDTO = _mapper.Map<IEnumerable<DataAccess.Entities.City>, IEnumerable<CityDTO>>(cities);
+
+            foreach (var city in citiesDTO)
+            {
+                city.Region.Administration = _mapper.Map<IEnumerable<RegionAdministration>, IEnumerable<RegionAdministrationDTO>>(cities.First(c => c.ID == city.ID).Region.RegionAdministration);
+            }
+            return citiesDTO;
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Admin/AdminService.cs:182]: Could you use linq here?

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs [hunk 3]
```diff
@@ -161,7 +119,82 @@ public async Task ChangeCurrentRoleAsync(string userId, string role)
...
+        {
+            var admin = _roleManager.Roles.Where(i => i.Name == "Admin");
+            var allRoles = await _roleManager.Roles.
+                Except(admin).
+                OrderBy(i => i.Name).
+                ToListAsync();
+            return allRoles;
+        }
+
+        public async Task UpdateUserDatesByChangeRole(string userId, string role)
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Admin/AdminService.cs:196]:
```suggestion
        public async Task UpdateUserDatesByChangeRoleAsync(string userId, string role)
```

File: EPlast/EPlast.BLL/Services/EmailSending/EmailReminderService.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                                    IMapper mapper,
+                                    UserManager<User> userManager)
+        {
+            _repoWrapper = repoWrapper;
+            _authEmailServices = authEmailServices;
+            _mapper = mapper;
+            _userManager = userManager;
+        }
+
+        public async Task JoinCityReminder()
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/EmailSending/EmailReminderService.cs:31]:
```suggestion
        public async Task JoinCityReminderAsync()
```

File: EPlast/EPlast.BLL/Services/EmailSending/EmailReminderService.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            var lonelyUsers = await GetLonelyUsersAsync();
+            foreach (var user in lonelyUsers)
+            {
+                await _authEmailServices.SendEmailJoinToCityReminderAsync(user.Email);
+            }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/EmailSending/EmailReminderService.cs:37]: Could you use linq here?

**Review by sooooolo — APPROVED** [review]

File: EPlast/EPlast.WebApi/Controllers/AuthController.cs [hunk 6]
```diff
@@ -13,8 +13,13 @@ namespace EPlast.WebApi.Controllers
...
     public class AuthController : ControllerBase
     {
+        private readonly IAuthEmailService _authEmailServices;
+        private readonly IAuthService _authService;
+        private readonly IHomeService _homeService;
+        private readonly IResources _resources;
+        private readonly IUserDatesService _userDatesService;
         public AuthController(
-            IAuthService authService,
+                                                    IAuthService authService,
```
Comment (voroniak) [inline EPlast/EPlast.WebApi/Controllers/AuthController.cs:22]: ?

File: EPlast/EPlast.WebApi/StartupExtensions/AddRecurringJobManager.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                                            "59 23 * * *", TimeZoneInfo.Local);
+            CreateRoles(serviceProvider, Configuration).Wait();
+            recurringJobManager.AddOrUpdate("Reminder to join city",
+                                            () => serviceProvider.GetService<IEmailReminderService>()
+                                                                 .JoinCityReminder(),
+                                            "0 12 * * Mon",
+                                            TimeZoneInfo.Local);
+        }
+
+        private static async Task CreateRoles(IServiceProvider serviceProvider, IConfiguration Configuration)
```
Comment (voroniak) [inline EPlast/EPlast.WebApi/StartupExtensions/AddRecurringJobManager.cs:56]:
```suggestion
        private static async Task CreateRolesAsync(IServiceProvider serviceProvider, IConfiguration Configuration)
```

**Review by voroniak — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs (hunk 2 above)
Comment (vovikboyko) [reply EPlast/EPlast.BLL/Services/Admin/AdminService.cs:182]: It's not my code. So, will add this as a new issue.

Comment (voroniak) [reply EPlast/EPlast.BLL/Services/Admin/AdminService.cs:182]: But you may fix this here

**Review by Lunaflumen — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs [hunk 8]
```diff
@@ -161,7 +119,79 @@ public async Task ChangeCurrentRoleAsync(string userId, string role)
...
+                city.Region.Administration = _mapper.Map<IEnumerable<RegionAdministration>, IEnumerable<RegionAdministrationDTO>>(cities.First(c => c.ID == city.ID).Region.RegionAdministration);
+            }
+            return citiesDTO;
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<IdentityRole> GetRolesExceptAdmin()
+        {
+            var admin = _roleManager.Roles.Where(i => i.Name == "Admin");
+            var allRoles = _roleManager.Roles.Except(admin).OrderBy(i => i.Name).ToList();
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Admin/AdminService.cs:189]: why do we have to se .ToList(); here ?

File: EPlast/EPlast.WebApi/Controllers/AdminController.cs [hunk 9]
```diff
@@ -232,10 +214,32 @@ public async Task<IActionResult> GetCityAndRegionAdminsOfUser(string userId)
...
+            return Ok(model);
+        }
+
+        /// <summary>
+        /// Get all users with additional information
+        /// </summary>
+        /// <returns>Specify model with all users</returns>
+        /// <response code="200">Successful operation</response>
+        [HttpGet("usersTable")]
+        public async Task<IActionResult> UsersTable()
```
Comment (voroniak) [inline EPlast/EPlast.WebApi/Controllers/AdminController.cs:239]:
```suggestion
        public async Task<IActionResult> GetUsersTable()
```

File: EPlast/EPlast.WebApi/Controllers/LoginController.cs [hunk 10]
```diff
@@ -146,13 +151,6 @@ public IActionResult Logout()
...
             return Ok();
         }
-
-        private readonly IAuthService _authService;
-        private readonly IJwtService _jwtService;
-        private readonly ILoggerService<LoginController> _loggerService;
-        private readonly IResources _resources;
-        private readonly IUserDatesService _userDatesService;
-
         private async Task AddEntryMembershipDate(string userId)
```
Comment (voroniak) [inline EPlast/EPlast.WebApi/Controllers/LoginController.cs:154]:
```suggestion
        private async Task AddEntryMembershipDateAsync(string userId)
```
Could you also change this ?

**Review by voroniak — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs (hunk 8 above)
Comment (vovikboyko) [reply EPlast/EPlast.BLL/Services/Admin/AdminService.cs:189]: fixed

File: EPlast/EPlast.WebApi/Controllers/AdminController.cs (hunk 9 above)
Comment (vovikboyko) [reply EPlast/EPlast.WebApi/Controllers/AdminController.cs:239]: done

File: EPlast/EPlast.WebApi/Controllers/LoginController.cs (hunk 10 above)
Comment (vovikboyko) [reply EPlast/EPlast.WebApi/Controllers/LoginController.cs:154]: done

**Review by voroniak — APPROVED** [review]
LGTM!

**Pull request merged**